src/serializers/type_serializers/union.rs

```diff
@@ -80,21 +131,16 @@ fn to_python(
     // try the serializers in left to right order with error_on fallback=true
     let mut new_extra = extra.clone();
     new_extra.check = SerCheck::Strict;
-    let mut errors: SmallVec<[PyErr; SMALL_UNION_THRESHOLD]> = SmallVec::new();
 
-    for comb_serializer in choices {
-        match comb_serializer.to_python(value, include, exclude, &new_extra) {
-            Ok(v) => return Ok(v),
-            Err(err) => errors.push(err),
-        }
-    }
+    let errors = match to_python_extractor(value, include, exclude, &new_extra, choices) {
+        ToPythonExtractorResult::Success(obj) => return Ok(obj),
+        ToPythonExtractorResult::Errors(errs) => errs,
+    };
 
     if retry_with_lax_check {
         new_extra.check = SerCheck::Lax;
-        for comb_serializer in choices {
-            if let Ok(v) = comb_serializer.to_python(value, include, exclude, &new_extra) {
-                return Ok(v);
-            }
+        if let ToPythonExtractorResult::Success(v) = to_python_extractor(value, include, exclude, &new_extra, choices) {
+            return Ok(v);
         }
     }
 
```

Comment:
We should check if `extra.check != SerCheck::None` here. That would tell us if we're inside a nested. union serialization. If so, we should return a `PydanticSerializationUnexpectedValue` error here containing the errors.

That way, the warning and the inference fallback below this point will only happen at the top level union.